### ISSUE TITLE
Memory leak only when function contains return statement

### ISSUE THREAD

**Issue by wouter89** [facebook/infer#336]
Hi,

When running Infer I came across some odd behaviour I couldn't explain. I have made some small code examples to illustrate the problem without having to deal with the context.

**Sample 1**:
For this code Infer doesn't report a memory leak, even though the condition in the if statement will always evaluate to false and data will never be freed. Infer does recognise that the condition is always false, so the problem doesn't lies with evaluation of the conditional statement.

```C
#include <wchar.h>

void memoryLeakingFunction()
{
	char * data;
	data = (char *)calloc(100, sizeof(char));

	if(0) //Never true
	{
		free(data);
	}
}

int main()
{
	memoryLeakingFunction();
	return 0;
}
```

**Sample 2** :
But, (strangely) Infer does report a memory leak for the following snippet of code.

```C
#include <wchar.h>

void memoryLeakingFunction()
{
	char * data;
	data = (char *)calloc(100, sizeof(char));

	if(0)  //Never true
	{
		free(data);
	}

	return;
}

int main()
{
	memoryLeakingFunction();
	return 0;
}
```

I can't figure out why the code in **sample 2** does report a memory leak. The only difference between the code snippets is the return statement. I would like for Infer to also report the memory leak in **sample 1**. 

Can you explain why Infer behaves this way and if I need to add return statements to my code?

**Comment by cristianoc** [comment]
Thank you for the report!

This seems to happen when the conditional is the last statement.
If you add another instruction, whether a return or something else, the leak will be reported.

We will fix this.

**Comment by cristianoc** [comment]
Fixed in 054ad223a83d34fe3b79cd7735f2e8e9e5738f44
